### PR TITLE
perf(hy_v4): add optional fused decode TopK transform

### PR DESCRIPTION
**Pull request by alexanderbin123**
## What changed

HY4 sparse-MLA decode can now call LightOp `fast_topk_transform_fused` on top of the latest MR 126 branch (including MR 130). The route preserves request-local indices with a cached identity page table, reuses stable unit-query cumulative lengths for CUDA Graph capture, copies into the existing TopK output buffer, and falls back to `top_k_per_row_decode` when the API is unavailable or `topk != 2048`.

Enable this route with:

```bash
export VLLM_HCU_USE_LIGHTOP_SPARSE_MLA_TOPK=1
export VLLM_HCU_USE_LIGHTOP_MASK_TOPK=0
export VLLM_HCU_USE_LIGHTOP_FAST_TOPK_TRANSFORM=1
```

`VLLM_HCU_USE_LIGHTOP_MASK_TOPK=0` is required for this A/B because the separate mask-TopK route has higher dispatch priority and bypasses the ordinary decode TopK helper when eligible.

The new flag defaults to `False`. On BW1100 at the requested low-latency shape (DP8 per-rank 16 requests, MTP3 with 4 decode rows per request, 64x8192 logits, TopK 2048), the fused identity-transform path was slower than the current direct-output kernel:

| Mode | Existing | Fused transform | Change |
|---|---:|---:|---:|
| Eager median | 54.96 us | 81.92 us | +49.0% latency |
| CUDA Graph median | 33.28 us | 45.12 us | +35.6% latency |

Keeping it opt-in avoids regressing MR 126's default path while making the requested LightOp integration available for further page-table work.

## Validation

- `194 passed`: LightOp API boundary, environment routing, and HY4 attention tests.
- HCU operator A/B: fused, existing LightOp, and PyTorch reference selected identical TopK sets for 64x8192 / TopK 2048 / MTP width 4.
- Independent code review: no remaining Critical, Important, or Minor findings.
- DP8 + EP8 DeepEP low-latency end-to-end passed with MTP3, FP8 KV dequant, CUDA Graph, and 16 scheduled requests per rank (aggregate cap 128). HumanEval limit-8 scored 8/8 (Accuracy and Pass@1 100%) with 8 successful requests and no service errors. MTP accepted 736/774 draft tokens (95.09%).

### CODE REVIEW

**Comment by alexanderbin123** [comment]
End-to-end validation completed on `fe0ab7c0` (2026-09-21).

Results:
- DP8 + EP8, `deepep_low_latency`, MTP3, FP8 KV dequant, CUDA Graph.
- `--max-num-seqs 16` per DP rank, aggregate scheduler cap 128.
- CUDA Graph capture completed for PIECEWISE/FULL prefill and decode.
- Smoke request: HTTP 200.
- HumanEval `--limit 8`: **8/8**, Accuracy **100%**, Pass@1 **100%**.
- Evaluation API requests: 8 succeeded, 0 errored.
- Mean latency: 6.921 s/request; average output throughput: 17.75 tok/s (batch size 1 accuracy run; not a controlled performance A/B).
- MTP3 metrics: 258 draft steps, 774 draft tokens, 736 accepted tokens, **95.09% acceptance**.
- Service log contained no traceback/runtime/engine errors.

Exact service launch used:

```bash
#!/usr/bin/env bash
set -euo pipefail
export PYTHONPATH=/tmp/vpd25-lightop-kv-reuse${PYTHONPATH:+:$PYTHONPATH}
export HIP_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export VLLM_USE_V2_MODEL_RUNNER=1
export VLLM_HCU_USE_CUSTOM_FLASH_ATTN=1
export VLLM_HCU_HYV4_FP8_KV_DEQUANT=1
export VLLM_HCU_USE_LIGHTOP_SPARSE_MLA_TOPK=1
export VLLM_HCU_USE_LIGHTOP_MASK_TOPK=0
export VLLM_HCU_USE_LIGHTOP_FAST_TOPK_TRANSFORM=1
export TORCH_BLAS_PREFER_CUBLASLT=1
export VLLM_DP_MASTER_IP=127.0.0.1
export VLLM_DP_MASTER_PORT=29641
unset VLLM_PLUGINS HTTP_PROXY HTTPS_PROXY ALL_PROXY http_proxy https_proxy all_proxy
export NO_PROXY=127.0.0.1,localhost
export no_proxy=127.0.0.1,localhost
exec python -m vllm.entrypoints.cli.main serve /models/Hy4-preview-Channel-FP8-w8a8 \
  --host 127.0.0.1 \
  --port 20131 \
  --served-model-name hy4-dp8-ep8-mtp3-fast-topk \
  --trust-remote-code \
  --dtype bfloat16 \
  -q compressed-tensors \
  --tensor-parallel-size 1 \
  --data-parallel-size 8 \
  --enable-expert-parallel \
  --all2all-backend deepep_low_latency \
  --moe-backend deep_gemm \
  --kv-cache-dtype fp8_ds_mla \
  --block-size 64 \
  --max-model-len 8192 \
  --max-num-seqs 16 \
  --max-num-batched-tokens 256 \
  --gpu-memory-utilization 0.95 \
  --kv-cache-memory-bytes 536870912 \
  --enable-prefix-caching \
  --default-chat-template-kwargs '{"reasoning_effort":"no_think"}' \
  --reasoning-parser hy_v4 \
  --enable-auto-tool-choice \
  --tool-call-parser hy_v4 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3}' \
  --cudagraph-capture-sizes 1 2 4 8 16 32 64 \
  --data-parallel-rpc-port 29642
```

The HumanEval command used:

```bash
env -u HTTP_PROXY -u HTTPS_PROXY -u ALL_PROXY -u http_proxy -u https_proxy -u all_proxy \
  NO_PROXY=127.0.0.1,localhost no_proxy=127.0.0.1,localhost \
  evalscope eval \
    --model hy4-dp8-ep8-mtp3-fast-topk \
    --api-url http://127.0.0.1:20131/v1 \
    --api-key EMPTY \
    --eval-type openai_api \
    --generation-config '{"temperature":0,"do_sample":false,"seed":0,"max_tokens":2048,"extra_body":{"chat_template_kwargs":{"reasoning_effort":"no_think"}}}' \
    --eval-batch-size 1 \
    --limit 8 \
    --datasets humaneval \
    --work-dir /tmp/vpd25-lightop-kv-reuse/_validation_fast_topk/e2e_humaneval8/eval_20260921 \
    --no-timestamp
```

**Comment by alexanderbin123** [comment]
已将 DeepEP dispatch 前 fused-gate expert IDs 的 int32→int64 转换消除叠加到本 MR，commit：`25baea5`。

实现说明：
- HY4 `num_expert_group=1` 且 DeepEP 请求 int64 IDs 时，若当前 LightOp 暴露 `output_indices_int64`，直接由 `moe_fused_gate` 输出 int64。
- 旧版 LightOp 或其他 grouped kernel 保持原有 ABI，并自动回退到调用后的 dtype 转换。
- 实际形状 `tokens=16, experts=256, topk=8` 的 10 次算子 profile：routing CUDA 总时间由 188.939 us 降至 119.830 us（-36.58%）；每次 5.679 us 的 `aten::to/copy_` elementwise kernel 已消失。该结果是 routing 微基准，不代表端到端同比收益。

验证环境：`deep-ep==1.1.0+dtk2604.torch2110.2609171757.gc807df`。

验证结果：
- `tests/runtime_patch/test_moe_deepep.py`：94 passed。
- `tests/runtime_patch/test_lightop_categorized_api.py`：10 passed。
- DP8 + EP8 + DeepEP low latency + MTP3 + CUDA Graph + FP8 KV dequant 服务启动正常。
- HumanEval limit-8：8/8，Accuracy 100%，Pass@1 100%。
- 服务日志未发现 traceback、runtime error 或 `invalid tokens_per_request`。

本次端到端验证使用的启动命令：

```bash
cd /models/vllm-plugin-das
export PYTHONPATH="$PWD${PYTHONPATH:+:$PYTHONPATH}"
export HIP_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export VLLM_USE_V2_MODEL_RUNNER=1
export VLLM_MXFP8_EMULATION_DEQUANT_AT_LOAD=0
export VLLM_HCU_USE_CUSTOM_OPS=1
export VLLM_HCU_USE_FUSE_MOE_GATE=1
export VLLM_HCU_USE_CUSTOM_FLASH_ATTN=1
export VLLM_HCU_HYV4_FP8_KV_DEQUANT=1
export VLLM_HCU_USE_LIGHTOP_SPARSE_MLA_TOPK=1
export VLLM_HCU_USE_LIGHTOP_MASK_TOPK=0
export VLLM_HCU_USE_LIGHTOP_FAST_TOPK_TRANSFORM=1
export VLLM_HCU_USE_AITER_OPUS_PAGED_MQA_LOGITS=1
export TORCH_BLAS_PREFER_CUBLASLT=1
export ROCSHMEM_HEAP_SIZE=4000000000
export VLLM_DP_MASTER_IP=127.0.0.1
export VLLM_DP_MASTER_PORT=29688
unset VLLM_PLUGINS HTTP_PROXY HTTPS_PROXY ALL_PROXY http_proxy https_proxy all_proxy
export NO_PROXY=127.0.0.1,localhost
export no_proxy=127.0.0.1,localhost
python -m vllm.entrypoints.cli.main serve /models/Hy4-preview-Channel-FP8-w8a8 \
  --host 127.0.0.1 \
  --port 20138 \
  --served-model-name hy4-dp8-ep8-mtp3-gate-probe \
  --trust-remote-code \
  --dtype bfloat16 \
  -q compressed-tensors \
  --tensor-parallel-size 1 \
  --data-parallel-size 8 \
  --enable-expert-parallel \
  --all2all-backend deepep_low_latency \
  --moe-backend deep_gemm \
  --kv-cache-dtype fp8_ds_mla \
  --block-size 64 \
  --max-model-len 8192 \
  --max-num-seqs 16 \
  --max-num-batched-tokens 256 \
  --gpu-memory-utilization 0.95 \
  --kv-cache-memory-bytes 536870912 \
  --enable-prefix-caching \
  --default-chat-template-kwargs '{"reasoning_effort":"no_think"}' \
  --reasoning-parser hy_v4 \
  --enable-auto-tool-choice \
  --tool-call-parser hy_v4 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3,"use_local_argmax_reduction":true}' \
  --cudagraph-capture-sizes 1 2 4 8 16 32 64 \
  --data-parallel-rpc-port 29689
```

**Comment by alexanderbin123** [comment]
已完成 AITER Opus `paged_mqa_logits` 路由修复，并按 MR 131 最新分支的完整配置做了端到端验证。

修复点：

- `743d36e`：兼容 MTP metadata 的二维 `context_lens`。AITER 调用前取每个请求最后一列并转成连续的一维 tensor。
- `e4bb5f7`：显式设置 `VLLM_HCU_USE_AITER_OPUS_PAGED_MQA_LOGITS=1` 时，即使全局 legacy AITER 已开启，也优先进入原生 wrapper，避免 Opus 路由被 legacy indexer 截走。

验证结果：

- 精确使用下方 DP8 + EP8 + MTP3 + FP8 KV dequant + CUDA Graph 命令启动成功，8 个 API server 全部就绪。
- CUDA Graph 的 base PIECEWISE/FULL 与 MTP prefill/decode graph 均完成捕获；约 17–18 秒、约 2.03 GiB/卡。
- 直接运行时探针确认 8 个 worker 均调用 AITER Opus：`q=(64, 1, 32, 128)`，`context_lens=(64, 1)` 转为 `(64,)`，`block_tables=(64, 128)`，`max_model_len=8192`。
- HumanEval 8 条：8/8，Accuracy 100%，Pass@1 100%，平均 latency 9.144 s，平均 throughput 14.67 tok/s。
- 回归测试：`143 passed`。
- 服务已停止，8 张卡已释放。

启动命令：

```bash
cd /models/vllm-plugin-das
export PYTHONPATH="$PWD${PYTHONPATH:+:$PYTHONPATH}"
export HIP_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export VLLM_USE_V2_MODEL_RUNNER=1
export VLLM_MXFP8_EMULATION_DEQUANT_AT_LOAD=0
export VLLM_HCU_USE_CUSTOM_OPS=1
export VLLM_HCU_USE_FUSE_MOE_GATE=1
export VLLM_HCU_USE_CUSTOM_FLASH_ATTN=1
export VLLM_HCU_HYV4_FP8_KV_DEQUANT=1
export VLLM_HCU_USE_LIGHTOP_SPARSE_MLA_TOPK=1
export VLLM_HCU_USE_LIGHTOP_MASK_TOPK=0
export VLLM_HCU_USE_LIGHTOP_FAST_TOPK_TRANSFORM=1
export VLLM_HCU_USE_AITER_OPUS_PAGED_MQA_LOGITS=1
export TORCH_BLAS_PREFER_CUBLASLT=1
export ROCSHMEM_HEAP_SIZE=4000000000
export VLLM_DP_MASTER_IP=127.0.0.1
export VLLM_DP_MASTER_PORT=29688
unset VLLM_PLUGINS HTTP_PROXY HTTPS_PROXY ALL_PROXY http_proxy https_proxy all_proxy
export NO_PROXY=127.0.0.1,localhost
export no_proxy=127.0.0.1,localhost
python -m vllm.entrypoints.cli.main serve /models/Hy4-preview-Channel-FP8-w8a8 \
  --host 127.0.0.1 \
  --port 20138 \
  --served-model-name hy4-dp8-ep8-mtp3-gate-probe \
  --trust-remote-code \
  --dtype bfloat16 \
  -q compressed-tensors \
  --tensor-parallel-size 1 \
  --data-parallel-size 8 \
  --enable-expert-parallel \
  --all2all-backend deepep_low_latency \
  --moe-backend deep_gemm \
  --kv-cache-dtype fp8_ds_mla \
  --block-size 64 \
  --max-model-len 8192 \
  --max-num-seqs 16 \
  --max-num-batched-tokens 256 \
  --gpu-memory-utilization 0.95 \
  --kv-cache-memory-bytes 536870912 \
  --enable-prefix-caching \
  --default-chat-template-kwargs '{"reasoning_effort":"no_think"}' \
  --reasoning-parser hy_v4 \
  --enable-auto-tool-choice \
  --tool-call-parser hy_v4 \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3,"use_local_argmax_reduction":true}' \
  --cudagraph-capture-sizes 1 2 4 8 16 32 64 \
  --data-parallel-rpc-port 29689
```

**Comment by alexanderbin123** [comment]
补充 AITER Opus `paged_mqa_logits` 性能 A/B 结果。

测试配置保持一致：DP8 + EP8 + MTP3、`deepep_low_latency`、FP8 KV dequant、全局并发 128（每卡 16）、固定输入 384 tokens / 输出 128 tokens、默认 CUDA Graph 策略（未传 `--compilation-config`，解析为 `FULL_AND_PIECEWISE`）、固定 seed；每组完整预热后正式测试 3 轮，以下取中位数。

| 指标 | AITER Opus ON | LightOp baseline | AITER 相对变化 |
|---|---:|---:|---:|
| 输出吞吐 | 392.76 tok/s | 409.58 tok/s | -4.11% |
| 请求吞吐 | 3.068 req/s | 3.200 req/s | -4.11% |
| 总吞吐 | 1571.03 tok/s | 1638.31 tok/s | -4.11% |
| 完成 128 请求耗时 | 41.72 s | 40.00 s | +4.28% |
| Mean TPOT | 159.93 ms | 158.86 ms | +0.68% |
| Mean ITL | 433.89 ms | 424.56 ms | +2.20% |
| MTP acceptance | 57.44% | 57.32% | 基本一致 |

对应 CUDA Graph 算子单测（context=512、heads=32、head_dim=128、max_model_len=8192）：

- 每卡实际 batch=16：AITER 14.47 us，LightOp 10.88 us；AITER 延迟高约 33.0%。
- batch=64：AITER 21.58 us，LightOp 24.86 us；AITER 延迟低约 13.2%。

结论：当前服务每卡最多处理 16 路，请求落在 AITER 的小 batch 劣势区间；开启 `VLLM_HCU_USE_AITER_OPUS_PAGED_MQA_LOGITS=1` 没有端到端收益，本次测量约回退 4.1%。建议继续保留 opt-in，DP8/每卡 16 的低延时配置默认不开。

测试中还发现：直接以 LightOp 路径启动时，memory profile 阶段两次复现 rocBLAS 缺失 Tensile module 的 `HIPBLAS_STATUS_INTERNAL_ERROR`。为完成可比基线，服务先用 AITER 完成 memory profile，随后 8/8 worker 在 CUDA Graph 捕获前切换到 LightOp；日志确认切换顺序和 LightOp 图捕获均完成。该启动问题与正式 A/B 数据分开记录。
